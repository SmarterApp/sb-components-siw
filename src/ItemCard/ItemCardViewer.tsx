import * as React from "react";
import {
  ItemViewerFrame,
  RubricTable,
  AboutThisItemDetail,
  ItemTabs,
  Tabs,
  AboutItemModel
} from "@src/index";

export interface ItemCardViewerProps {
  item?: AboutItemModel;
}

export interface State {
  selectedTab: Tabs;
}

export class ItemCardViewer extends React.Component<
  ItemCardViewerProps,
  State
> {
  constructor(props: ItemCardViewerProps) {
    super(props);
    this.state = {
      selectedTab: "viewer"
    };
  }

  onTabChange(tab: Tabs) {
    this.setState({
      selectedTab: tab
    });
  }

  renderViewer(url: string) {
    return (
      <div className="item-content">
        <ItemViewerFrame url={url} />
      </div>
    );
  }

  renderRubric() {
    const aboutItem = this.props.item;

    if (aboutItem && aboutItem.sampleItemScoring) {
      if (aboutItem.sampleItemScoring.rubrics) {
        return (
          <div className="item-content">
            <RubricTable rubrics={aboutItem.sampleItemScoring.rubrics} />
          </div>
        );
      }
    }
  }

  renderInformation() {
    if (this.props.item) {
      const aboutItem = this.props.item;

      return (
        <div className="item-content">
          <div>
            <AboutThisItemDetail {...aboutItem} />
          </div>
        </div>
      );
    }
  }

  renderChosen() {
    if (this.props.item) {
      const selectedTab = this.state.selectedTab;
      const itemCard = this.props.item.itemCardViewModel;

      let resultElement: JSX.Element[] | JSX.Element | undefined;
      if (selectedTab === "viewer") {
        const url = `http://ivs.smarterbalanced.org/items?ids=${
          itemCard.bankKey
        }-${itemCard.itemKey}`;
        resultElement = <div> {this.renderViewer(url)} </div>;
      } else if (selectedTab === "rubric") {
        resultElement = <div> {this.renderRubric()} </div>;
      } else if (selectedTab === "information") {
        resultElement = <div> {this.renderInformation()}</div>;
      }

      return resultElement;
    }
  }
  render() {
    let rubricVisibility = false;
    if (
      this.props.item &&
      this.props.item.sampleItemScoring &&
      this.props.item.sampleItemScoring.rubrics
    ) {
      rubricVisibility = true;
    }

    return (
      <div className="item-card">
        <ItemTabs
          showRubricTab={rubricVisibility}
          changedTab={tab => this.onTabChange(tab)}
          selectedTab={this.state.selectedTab}
        />
        {this.renderChosen()}
      </div>
    );
  }
}
