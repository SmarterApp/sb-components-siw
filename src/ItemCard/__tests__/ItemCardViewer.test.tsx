import * as React from "react";
import * as ReactDOM from "react-dom";
import { ItemCardViewer } from "../ItemCardViewer";
import * as TestUtils from "react-dom/test-utils";
import { shallow, mount, render } from "enzyme";

import { ItemCardModel, RubricModel, AboutItemModel } from "@src/index";
import { aboutItemMockModel } from "@mocks/index";

describe("ItemCardViewer", () => {
  it("renders viewer tab", () => {
    const wrapper = shallow(<ItemCardViewer item={aboutItemMockModel} />);
    wrapper.setState({ selectedTab: "viewer" });
    expect(wrapper).toMatchSnapshot();
  });

  it("renders rubric tab", () => {
    const wrapper = shallow(<ItemCardViewer item={aboutItemMockModel} />);
    wrapper.setState({ selectedTab: "rubric" });
    expect(wrapper).toMatchSnapshot();
  });

  it("renders information tab", () => {
    const wrapper = shallow(<ItemCardViewer item={aboutItemMockModel} />);
    wrapper.setState({ selectedTab: "information" });
    expect(wrapper).toMatchSnapshot();
  });
});
