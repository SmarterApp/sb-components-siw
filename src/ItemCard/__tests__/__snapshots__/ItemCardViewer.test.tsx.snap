// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCardViewer renders information tab 1`] = `
<div
  className="item-card"
>
  <ItemTabs
    changedTab={[Function]}
    selectedTab="information"
    showRubricTab={true}
  />
  <div>
     
    <div
      className="item-content"
    >
      <div>
        <AboutThisItemDetail
          commonCoreStandardsDescription="A-123-BC"
          depthOfKnowledge="3"
          educationalDifficulty="5"
          evidenceStatement="This is a sample evidence statement"
          itemCardViewModel={
            Object {
              "answerKeys": "",
              "availableBrailleTypes": Array [
                Object {
                  "disabled": false,
                  "hidden": false,
                  "label": "EBAE-Contracted",
                  "selectionCode": "EABE",
                },
                Object {
                  "disabled": false,
                  "hidden": false,
                  "label": "BAET-Contracted",
                  "selectionCode": "BAET",
                },
              ],
              "bankKey": 187,
              "brailleOnlyItem": false,
              "calculator": false,
              "ccssDescription": "this is the common core description",
              "claimCode": "ELA3",
              "claimLabel": "Claim",
              "commonCoreStandardId": "AB-123",
              "depthOfKnowledge": "2",
              "domain": "domain",
              "grade": 4,
              "gradeLabel": "Grade 4",
              "interactionTypeCode": "GI",
              "interactionTypeLabel": "Item Type",
              "isPerformanceItem": false,
              "itemDifficulty": "Hard",
              "itemKey": 3434,
              "selected": false,
              "subjectCode": "ELA",
              "subjectLabel": "ELA",
              "targetDescription": "target description right here",
              "targetHash": 33434,
              "targetId": "dfas",
              "targetShortName": "target short name",
              "title": "ELA Grade 4",
            }
          }
          sampleItemScoring={
            Object {
              "answerKey": "x = 5",
              "hasMachineRubric": false,
              "rubrics": Array [
                Object {
                  "language": "en-us",
                  "rubricEntries": Array [
                    Object {
                      "name": "No points",
                      "scorepoint": "0",
                      "value": "Student did not do the problem",
                    },
                    Object {
                      "name": "Half points",
                      "scorepoint": "1",
                      "value": "Student tried to do the problem, but got it wrong",
                    },
                    Object {
                      "name": "No points",
                      "scorepoint": "2",
                      "value": "Student did the problem correctly",
                    },
                  ],
                  "samples": Array [
                    Object {
                      "maxValue": "0",
                      "minValue": "0",
                      "sampleResponses": Array [
                        Object {
                          "name": "0pts",
                          "purpose": "idk",
                          "sampleContent": "x = _",
                          "scorePoint": "0",
                        },
                      ],
                    },
                    Object {
                      "maxValue": "1",
                      "minValue": "1",
                      "sampleResponses": Array [
                        Object {
                          "name": "1pt",
                          "purpose": "idk",
                          "sampleContent": "x = 10 + 5 = 15",
                          "scorePoint": "1",
                        },
                        Object {
                          "name": "1pt",
                          "purpose": "idk",
                          "sampleContent": "x = 10 - 5 = a million",
                          "scorePoint": "1",
                        },
                      ],
                    },
                    Object {
                      "maxValue": "2",
                      "minValue": "2",
                      "sampleResponses": Array [
                        Object {
                          "name": "2pts",
                          "purpose": "idk",
                          "sampleContent": "x = 10 - 5 = 5",
                          "scorePoint": "2",
                        },
                      ],
                    },
                  ],
                },
              ],
              "scoringOptions": Array [],
            }
          }
          targetDescription="Solve equations"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ItemCardViewer renders rubric tab 1`] = `
<div
  className="item-card"
>
  <ItemTabs
    changedTab={[Function]}
    selectedTab="rubric"
    showRubricTab={true}
  />
  <div>
     
    <div
      className="item-content"
    >
      <RubricTable
        rubrics={
          Array [
            Object {
              "language": "en-us",
              "rubricEntries": Array [
                Object {
                  "name": "No points",
                  "scorepoint": "0",
                  "value": "Student did not do the problem",
                },
                Object {
                  "name": "Half points",
                  "scorepoint": "1",
                  "value": "Student tried to do the problem, but got it wrong",
                },
                Object {
                  "name": "No points",
                  "scorepoint": "2",
                  "value": "Student did the problem correctly",
                },
              ],
              "samples": Array [
                Object {
                  "maxValue": "0",
                  "minValue": "0",
                  "sampleResponses": Array [
                    Object {
                      "name": "0pts",
                      "purpose": "idk",
                      "sampleContent": "x = _",
                      "scorePoint": "0",
                    },
                  ],
                },
                Object {
                  "maxValue": "1",
                  "minValue": "1",
                  "sampleResponses": Array [
                    Object {
                      "name": "1pt",
                      "purpose": "idk",
                      "sampleContent": "x = 10 + 5 = 15",
                      "scorePoint": "1",
                    },
                    Object {
                      "name": "1pt",
                      "purpose": "idk",
                      "sampleContent": "x = 10 - 5 = a million",
                      "scorePoint": "1",
                    },
                  ],
                },
                Object {
                  "maxValue": "2",
                  "minValue": "2",
                  "sampleResponses": Array [
                    Object {
                      "name": "2pts",
                      "purpose": "idk",
                      "sampleContent": "x = 10 - 5 = 5",
                      "scorePoint": "2",
                    },
                  ],
                },
              ],
            },
          ]
        }
      />
    </div>
     
  </div>
</div>
`;

exports[`ItemCardViewer renders viewer tab 1`] = `
<div
  className="item-card"
>
  <ItemTabs
    changedTab={[Function]}
    selectedTab="viewer"
    showRubricTab={true}
  />
  <div>
     
    <div
      className="item-content"
    >
      <ItemViewerFrame
        url="http://ivs.smarterbalanced.org/items?ids=187-3434"
      />
    </div>
     
  </div>
</div>
`;
